# Market Research Data Sources

## Industry Reports & Analytics

In [4]:
# Google Trends for woodworking demand
from pytrends.request import TrendReq

def analyze_woodworking_trends():
    pytrends = TrendReq(hl='en-US', tz=360)
    
    # Woodworking-related keywords
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
    
    # Get interest over time
    trends_data = pytrends.interest_over_time()
    
    # Get related queries
    related_queries = pytrends.related_queries()
    
    return trends_data, related_queries

trends, queries = analyze_woodworking_trends()

/home/lubuntu/myenv/lib/python3.12/site-packages/pytrends/request.py:260: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.fillna(False)


IndexError: list index out of range

### Anthropic Attempted Fix

In [2]:
import time
import random
from pytrends.request import TrendReq
import pandas as pd

def analyze_woodworking_trends():
    try:
        # Initialize with different settings
        pytrends = TrendReq(hl='en-US', tz=360, timeout=(10,25), retries=2, backoff_factor=0.1)
        
        # Use fewer keywords (max 5 per request)
        keywords = ['custom furniture', 'woodworking', 'handmade furniture']
        
        # Add random delay to avoid rate limiting
        time.sleep(random.uniform(1, 3))
        
        print("Building payload...")
        pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
        
        print("Getting interest over time...")
        trends_data = pytrends.interest_over_time()
        
        # Add delay before next request
        time.sleep(random.uniform(2, 4))
        
        print("Getting related queries...")
        related_queries = pytrends.related_queries()
        
        return trends_data, related_queries
        
    except Exception as e:
        print(f"Error: {e}")
        print("Trying alternative approach...")
        return analyze_with_smaller_batches()

def analyze_with_smaller_batches():
    """Alternative approach with individual keyword analysis"""
    pytrends = TrendReq(hl='en-US', tz=360, timeout=(10,25), retries=2, backoff_factor=0.1)
    
    keywords = ['woodworking', 'custom furniture', 'handmade furniture']
    all_data = {}
    
    for keyword in keywords:
        try:
            print(f"Analyzing: {keyword}")
            time.sleep(random.uniform(2, 4))  # Longer delay between requests
            
            pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
            trends = pytrends.interest_over_time()
            
            if not trends.empty:
                all_data[keyword] = trends[keyword]
            
        except Exception as e:
            print(f"Failed for {keyword}: {e}")
            continue
    
    if all_data:
        combined_df = pd.DataFrame(all_data)
        return combined_df, None
    else:
        return None, None

# Run the analysis
print("Starting woodworking trends analysis...")
trends, queries = analyze_woodworking_trends()

if trends is not None and not trends.empty:
    print("\n=== TRENDS DATA ===")
    print(trends.head())
    print(f"\nShape: {trends.shape}")
    
    if queries:
        print("\n=== RELATED QUERIES ===")
        for keyword, query_data in queries.items():
            print(f"\n{keyword.upper()}:")
            if query_data['top'] is not None:
                print("Top queries:")
                print(query_data['top'].head())
else:
    print("No data retrieved. Google Trends might be blocking requests.")
    print("Try again later or use a different approach.")

Starting woodworking trends analysis...
Building payload...
Error: Retry.__init__() got an unexpected keyword argument 'method_whitelist'
Trying alternative approach...
Analyzing: woodworking
Failed for woodworking: Retry.__init__() got an unexpected keyword argument 'method_whitelist'
Analyzing: custom furniture
Failed for custom furniture: Retry.__init__() got an unexpected keyword argument 'method_whitelist'
Analyzing: handmade furniture
Failed for handmade furniture: Retry.__init__() got an unexpected keyword argument 'method_whitelist'
No data retrieved. Google Trends might be blocking requests.
Try again later or use a different approach.


### Google Fix

In [5]:
# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time # Import the time module to pause between requests

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    Fetches interest over time for a list of keywords and then fetches
    related queries for each keyword individually.

    Returns:
        pd.DataFrame: A DataFrame containing interest over time data.
        dict: A dictionary containing related queries for each keyword.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    # Woodworking-related keywords
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time for all keywords (for comparison) ---
    print("Fetching interest over time for all keywords...")
    pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
    trends_data = pytrends.interest_over_time()
    
    # --- 2. Get Related Queries for each keyword individually ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        # Build a new payload for each individual keyword
        pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
        
        # Add a small delay to avoid making requests too quickly
        time.sleep(1) 
        
        try:
            related_queries_dict[keyword] = pytrends.related_queries()
        except Exception as e:
            print(f"Could not fetch related queries for '{keyword}'. Error: {e}")
            related_queries_dict[keyword] = None

    return trends_data, related_queries_dict

# Run the analysis
trends_df, related_queries = analyze_woodworking_trends()

# --- Display the results ---
print("\n--- Interest Over Time ---")
print(trends_df.tail()) # Print the last 5 rows of the trend data

print("\n\n--- Related Queries ---")
for keyword, queries in related_queries.items():
    if queries:
        print(f"\nTop Related Queries for '{keyword}':")
        # The result is a dictionary where the value is another dictionary with 'top' and 'rising'
        top_queries = queries[keyword]['top']
        if top_queries is not None and not top_queries.empty:
            print(top_queries)
        else:
            print("No top related queries found.")

Fetching interest over time for all keywords...


/home/lubuntu/myenv/lib/python3.12/site-packages/pytrends/request.py:260: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.fillna(False)



Fetching related queries for each keyword...
Fetching for: 'custom furniture'
Could not fetch related queries for 'custom furniture'. Error: list index out of range
Fetching for: 'woodworking'
Could not fetch related queries for 'woodworking'. Error: list index out of range
Fetching for: 'handmade furniture'
Could not fetch related queries for 'handmade furniture'. Error: list index out of range
Fetching for: 'live edge table'
Could not fetch related queries for 'live edge table'. Error: list index out of range
Fetching for: 'custom cabinets'
Could not fetch related queries for 'custom cabinets'. Error: list index out of range

--- Interest Over Time ---
            custom furniture  woodworking  handmade furniture  \
date                                                            
2025-06-01                10           70                   1   
2025-06-08                12           68                   1   
2025-06-15                10           71                   1   
2025-06-22 

#### Google Response to Broken API Endpoint

     ```

# Google Trends for woodworking demand

from pytrends.request import TrendReq


def analyze_woodworking_trends():

pytrends = TrendReq(hl='en-US', tz=360)

# Woodworking-related keywords

keywords = ['custom furniture', 'woodworking', 'handmade furniture',

'live edge table', 'custom cabinets']

pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')

# Get interest over time

trends_data = pytrends.interest_over_time()

# Get related queries

related_queries = pytrends.related_queries()

return trends_data, related_queries


trends, queries = analyze_woodworking_trends()

```


/home/lubuntu/myenv/lib/python3.12/site-packages/pytrends/request.py:260: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`

df = df.fillna(False)


---------------------------------------------------------------------------

IndexError Traceback (most recent call last)

Cell In[4], line 21

17 related_queries = pytrends.related_queries()

19 return trends_data, related_queries

---> 21 trends, queries = analyze_woodworking_trends()

Cell In[4], line 17, in analyze_woodworking_trends()

14 trends_data = pytrends.interest_over_time()

16 # Get related queries

---> 17 related_queries = pytrends.related_queries()

19 return trends_data, related_queries

File ~/myenv/lib/python3.12/site-packages/pytrends/request.py:450, in TrendReq.related_queries(self)

447 # top queries

448 try:

449 top_df = pd.DataFrame(

--> 450 req_json['default']['rankedList'][0]['rankedKeyword'])

451 top_df = top_df[['query', 'value']]

452 except KeyError:

453 # in case no top queries are found, the lines above will throw a KeyError

IndexError: list index out of range

Outline

    Diagnosis of the IndexError: An explanation of why the IndexError: list index out of range occurs when using related_queries() with multiple keywords.

    The Solution: A description of the correct approach, which involves iterating through each keyword to fetch its related queries individually.

    Corrected Python Code: A complete, runnable code snippet that implements the solution and correctly fetches both interest over time for all keywords and related queries for each keyword.

    Explanation of Changes: A step-by-step breakdown of the modifications made to the original code.

    A Note on the FutureWarning: A brief explanation of the warning message appearing in the output.

1. Diagnosis of the IndexError

Hello! I can help you with this Python error. The IndexError: list index out of range you're encountering is a common issue when using the pytrends library.

The root cause of the problem is that the pytrends.related_queries() method is designed to fetch data for only one keyword at a time.

In your original code, you build a single payload with a list of five keywords:
pytrends.build_payload(keywords, ...)

While interest_over_time() can handle multiple keywords to generate a comparison chart, related_queries() cannot. When it receives a payload with more than one keyword, the Google Trends API returns a data structure that pytrends doesn't expect. Specifically, the list that should contain the related query data (rankedList) comes back empty. The library then tries to access the first element of this empty list ([0]), which triggers the IndexError.

2. The Solution

To fix this, you need to separate the logic for fetching "interest over time" and "related queries."

    Interest Over Time: You can continue to fetch this data in a single call with all your keywords to get a comparison.

    Related Queries: You must loop through your list of keywords and make a separate API call for each one. You can store the results in a dictionary where the keys are your keywords and the values are the corresponding related queries.

3. Corrected Python Code

Here is the corrected and complete code. This version first gets the comparison data for all keywords and then iterates through each keyword to safely get its related queries.
Python

# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time # Import the time module to pause between requests

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    Fetches interest over time for a list of keywords and then fetches
    related queries for each keyword individually.

    Returns:
        pd.DataFrame: A DataFrame containing interest over time data.
        dict: A dictionary containing related queries for each keyword.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    # Woodworking-related keywords
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time for all keywords (for comparison) ---
    print("Fetching interest over time for all keywords...")
    pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
    trends_data = pytrends.interest_over_time()
    
    # --- 2. Get Related Queries for each keyword individually ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        # Build a new payload for each individual keyword
        pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
        
        # Add a small delay to avoid making requests too quickly
        time.sleep(1) 
        
        try:
            related_queries_dict[keyword] = pytrends.related_queries()
        except Exception as e:
            print(f"Could not fetch related queries for '{keyword}'. Error: {e}")
            related_queries_dict[keyword] = None

    return trends_data, related_queries_dict

# Run the analysis
trends_df, related_queries = analyze_woodworking_trends()

# --- Display the results ---
print("\n--- Interest Over Time ---")
print(trends_df.tail()) # Print the last 5 rows of the trend data

print("\n\n--- Related Queries ---")
for keyword, queries in related_queries.items():
    if queries:
        print(f"\nTop Related Queries for '{keyword}':")
        # The result is a dictionary where the value is another dictionary with 'top' and 'rising'
        top_queries = queries[keyword]['top']
        if top_queries is not None and not top_queries.empty:
            print(top_queries)
        else:
            print("No top related queries found.")

4. Explanation of Changes

    Separate API Calls: The core logic is now split. The call to get interest_over_time remains as it was, using the full list of keywords.

    Loop for Related Queries: A for loop iterates through your keywords list.

    Single-Keyword Payload: Inside the loop, pytrends.build_payload([keyword], ...) is called. Note that keyword is enclosed in square brackets [] because the method expects a list, even if it's just a list containing one item.

    Storing Results: The results from pytrends.related_queries() are stored in a dictionary called related_queries_dict. This is a clean way to associate each keyword with its specific results.

    Error Handling & Delay:

        time.sleep(1): A one-second pause is added in the loop. It's good practice to add a small delay between API calls to avoid being rate-limited or blocked by Google for making too many requests too quickly.

        try...except: I've wrapped the related_queries() call in a try...except block. This makes the code more robust. If Google Trends has no related query data for a specific term, it might still cause an error, and this block will catch it, print a message, and allow the script to continue to the next keyword instead of crashing.

5. A Note on the FutureWarning

You may still see this warning in your output:
FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated...

This warning is not causing your IndexError. It originates from the pandas library, which pytrends uses internally. It's simply informing you that a function pandas uses will behave differently in a future version. This is a message for the developers of pytrends to address in a future update. You can safely ignore this warning as it does not affect the execution or results of your code.


Fetching related queries for each keyword...

Fetching for: 'custom furniture'

Could not fetch related queries for 'custom furniture'. Error: list index out of range

Fetching for: 'woodworking'

Could not fetch related queries for 'woodworking'. Error: list index out of range

Fetching for: 'handmade furniture'

Could not fetch related queries for 'handmade furniture'. Error: list index out of range

Fetching for: 'live edge table'

Could not fetch related queries for 'live edge table'. Error: list index out of range

Fetching for: 'custom cabinets'

Could not fetch related queries for 'custom cabinets'. Error: list index out of range


--- Interest Over Time ---

custom furniture woodworking handmade furniture \

date

2025-06-01 10 70 1

2025-06-08 12 68 1

2025-06-15 10 71 1

2025-06-22 10 73 1

2025-06-29 11 66 1


live edge table custom cabinets isPartial

date

2025-06-01 5 15 False

2025-06-08 4 15 False

2025-06-15 5 15 False

2025-06-22 5 15 False

2025-06-29 5 13 True



--- Related Queries ---


You've run into a frustrating and unfortunately common issue with the pytrends library. Your new output confirms that the error isn't just about sending multiple keywords at once, but is a more fundamental problem with the related_queries() function itself.

Let's break down what's happening and how to build a more robust script to handle it.

Outline

    Diagnosis of the Persistent Error: An explanation of why the IndexError continues to happen even with single keywords, pointing to an issue within the pytrends library itself.

    The Fragility of Unofficial Wrappers: A brief look at why libraries like pytrends can sometimes break.

    Robust Solution: A revised code block that anticipates these errors and safely handles the data that is returned, preventing crashes and extracting whatever information is available.

    Interpreting the New Results: How to understand the output from the robust script.

1. Diagnosis of the Persistent Error

The fact that you are now getting the IndexError: list index out of range for every single keyword tells us the problem lies within the pytrends library's ability to parse the response it's getting from Google.

Here's the situation:

    You send a valid request for a single keyword (e.g., 'woodworking').

    Google Trends sends back a response.

    The structure of this response has likely changed slightly from what the pytrends library was originally designed to expect.

    When the related_queries() method tries to find the data in the specific location it's hard-coded to look (req_json['default']['rankedList'][0]), it finds an empty list ([]) and crashes when it tries to get the first item ([0]).

This happens even when there are related queries visible on the Google Trends website. The issue is purely in the parsing step within the library.

2. The Fragility of Unofficial Wrappers

pytrends is an unofficial "wrapper" library. It works by mimicking a web browser to scrape the Google Trends website and access its internal API. This is a brilliant but fragile approach. If Google changes the backend structure of their website or API—even slightly—the library can break until its developers release a new version that accounts for the change. You appear to have caught it during one of these broken periods.

3. Robust Solution

We cannot fix the pytrends library itself, but we can write our Python code to be more defensive. The goal is to successfully call the function and then carefully inspect whatever data we get back, rather than assuming it's in the correct format.

The code below is designed to handle this IndexError gracefully. It attempts the call and then safely checks the structure of the returned dictionary before trying to extract the 'top' and 'rising' queries.
Python

# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    This robust version anticipates potential errors from the pytrends library
    and handles them gracefully.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time (this part usually works reliably) ---
    print("Fetching interest over time for all keywords...")
    try:
        pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
        trends_data = pytrends.interest_over_time()
    except Exception as e:
        print(f"Could not fetch interest over time. Error: {e}")
        trends_data = None
    
    # --- 2. Get Related Queries with Robust Error Handling ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        time.sleep(2) # Increased delay to be safe
        
        try:
            # Re-build payload for each keyword
            pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
            related_queries_dict[keyword] = pytrends.related_queries()
        except IndexError:
            # This is the specific error we are seeing.
            print(f"  > Caught IndexError. Google's API response is not in the expected format for '{keyword}'.")
            related_queries_dict[keyword] = None # Mark as failed
        except Exception as e:
            # Catch any other potential errors
            print(f"  > An unexpected error occurred for '{keyword}': {e}")
            related_queries_dict[keyword] = None # Mark as failed
            
    return trends_data, related_queries_dict

# --- Run the analysis and display results safely ---
trends_df, related_queries = analyze_woodworking_trends()

if trends_df is not None:
    print("\n\n--- Interest Over Time ---")
    print(trends_df.tail())
else:
    print("\n\n--- No Interest Over Time data was fetched ---")

print("\n\n--- Related Queries ---")
if not related_queries:
    print("No related queries could be fetched.")
else:
    for keyword, result in related_queries.items():
        print(f"\n--- Results for '{keyword}' ---")
        if result is None:
            print("  > No data returned due to an API parsing error.")
            continue
        
        # Safely access 'top' and 'rising' queries
        top_queries = result.get(keyword, {}).get('top')
        rising_queries = result.get(keyword, {}).get('rising')
        
        if top_queries is not None and not top_queries.empty:
            print("\n  Top Related Queries:")
            print(top_queries)
        else:
            print("\n  > No 'Top' related queries were found or could be parsed.")
            
        if rising_queries is not None and not rising_queries.empty:
            print("\n  Rising Related Queries:")
            print(rising_queries)
        else:
            print("\n  > No 'Rising' related queries were found or could be parsed.")

4. Interpreting the New Results

When you run this new script, you will likely see a similar output to what you got before, but with more descriptive messages:
Plaintext

Fetching for: 'custom furniture'
  > Caught IndexError. Google's API response is not in the expected format for 'custom furniture'.
... (and so on for other keywords)

At the end, the results section will clearly state that no data could be parsed.

What to do now?

    Wait for an Update: The most common solution is to wait for the pytrends developers to patch the library. You can check for updates periodically by running pip install --upgrade pytrends.

    Try Different Parameters: You can experiment by changing the timeframe. Sometimes a shorter timeframe like 'today 3-m' or 'now 7-d' might use a different API endpoint that isn't broken.

    Use the Website: For the time being, the most reliable way to get related queries is to go directly to the Google Trends website and search for your terms manually.



## E-commerce Platform APIs

In [6]:
# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    This robust version anticipates potential errors from the pytrends library
    and handles them gracefully.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time (this part usually works reliably) ---
    print("Fetching interest over time for all keywords...")
    try:
        pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
        trends_data = pytrends.interest_over_time()
    except Exception as e:
        print(f"Could not fetch interest over time. Error: {e}")
        trends_data = None
    
    # --- 2. Get Related Queries with Robust Error Handling ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        time.sleep(2) # Increased delay to be safe
        
        try:
            # Re-build payload for each keyword
            pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
            related_queries_dict[keyword] = pytrends.related_queries()
        except IndexError:
            # This is the specific error we are seeing.
            print(f"  > Caught IndexError. Google's API response is not in the expected format for '{keyword}'.")
            related_queries_dict[keyword] = None # Mark as failed
        except Exception as e:
            # Catch any other potential errors
            print(f"  > An unexpected error occurred for '{keyword}': {e}")
            related_queries_dict[keyword] = None # Mark as failed
            
    return trends_data, related_queries_dict

# --- Run the analysis and display results safely ---
trends_df, related_queries = analyze_woodworking_trends()

if trends_df is not None:
    print("\n\n--- Interest Over Time ---")
    print(trends_df.tail())
else:
    print("\n\n--- No Interest Over Time data was fetched ---")

print("\n\n--- Related Queries ---")
if not related_queries:
    print("No related queries could be fetched.")
else:
    for keyword, result in related_queries.items():
        print(f"\n--- Results for '{keyword}' ---")
        if result is None:
            print("  > No data returned due to an API parsing error.")
            continue
        
        # Safely access 'top' and 'rising' queries
        top_queries = result.get(keyword, {}).get('top')
        rising_queries = result.get(keyword, {}).get('rising')
        
        if top_queries is not None and not top_queries.empty:
            print("\n  Top Related Queries:")
            print(top_queries)
        else:
            print("\n  > No 'Top' related queries were found or could be parsed.")
            
        if rising_queries is not None and not rising_queries.empty:
            print("\n  Rising Related Queries:")
            print(rising_queries)
        else:
            print("\n  > No 'Rising' related queries were found or could be parsed.")

Fetching interest over time for all keywords...


/home/lubuntu/myenv/lib/python3.12/site-packages/pytrends/request.py:260: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.fillna(False)



Fetching related queries for each keyword...
Fetching for: 'custom furniture'
  > Caught IndexError. Google's API response is not in the expected format for 'custom furniture'.
Fetching for: 'woodworking'
  > Caught IndexError. Google's API response is not in the expected format for 'woodworking'.
Fetching for: 'handmade furniture'
  > Caught IndexError. Google's API response is not in the expected format for 'handmade furniture'.
Fetching for: 'live edge table'
  > Caught IndexError. Google's API response is not in the expected format for 'live edge table'.
Fetching for: 'custom cabinets'
  > Caught IndexError. Google's API response is not in the expected format for 'custom cabinets'.


--- Interest Over Time ---
            custom furniture  woodworking  handmade furniture  \
date                                                            
2025-06-01                10           70                   1   
2025-06-08                12           68                   1   
2025-06-15     

In [ ]:
# Etsy API for handmade/custom furniture market
import requests

def analyze_etsy_woodworking_market(api_key):
    """Analyze woodworking market on Etsy"""
    
    # Search for woodworking items
    url = "https://openapi.etsy.com/v3/application/listings/active"
    
    headers = {
        'x-api-key': api_key,
        'Content-Type': 'application/json'
    }
    
    params = {
        'keywords': 'custom furniture woodworking',
        'limit': 100,
        'includes': 'Images,Shop,User',
        'sort_on': 'score'
    }
    
    response = requests.get(url, headers=headers, params=params)
    listings = response.json()
    
    # Analyze pricing and demand indicators
    market_data = []
    for listing in listings.get('results', []):
        market_data.append({
            'title': listing.get('title'),
            'price': listing.get('price', {}).get('amount'),
            'currency': listing.get('price', {}).get('currency_code'),
            'views': listing.get('views'),
            'num_favorers': listing.get('num_favorers'),
            'shop_name': listing.get('Shop', {}).get('shop_name'),
            'creation_timestamp': listing.get('creation_timestamp')
        })
    
    return market_data

## Local Business Intelligence

In [1]:
# Google My Business API for local competition analysis
def analyze_local_woodworking_businesses(location, gmb_api_key):
    """Find local woodworking businesses and their performance"""
    import googlemaps
    
    gmaps = googlemaps.Client(key=gmb_api_key)
    
    # Search for woodworking businesses
    search_terms = [
        'custom furniture',
        'woodworking shop',
        'cabinet maker',
        'furniture maker'
    ]
    
    local_businesses = []
    
    for term in search_terms:
        places = gmaps.places_nearby(
            location=location,
            radius=50000,  # 50km radius
            keyword=term,
            type='establishment'
        )
        
        for place in places.get('results', []):
            place_details = gmaps.place(
                place_id=place['place_id'],
                fields=['name', 'rating', 'user_ratings_total', 'website', 
                       'formatted_phone_number', 'opening_hours', 'photos']
            )
            
            local_businesses.append({
                'name': place_details['result']['name'],
                'rating': place_details['result'].get('rating'),
                'review_count': place_details['result'].get('user_ratings_total'),
                'website': place_details['result'].get('website'),
                'search_term': term
            })
    
    return local_businesses

## Price Analysis Tools

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

def analyze_pricing_data(market_data):
    """Analyze pricing patterns in woodworking market"""
    
    df = pd.DataFrame(market_data)
    
    # Price analysis by category
    price_analysis = {
        'avg_price': df['price'].mean(),
        'median_price': df['price'].median(),
        'price_range': {
            'low': df['price'].quantile(0.25),
            'high': df['price'].quantile(0.75)
        },
        'popular_price_points': df['price'].value_counts().head(10)
    }
    
    # Demand indicators
    demand_analysis = {
        'avg_views': df['views'].mean(),
        'avg_favorites': df['num_favorers'].mean(),
        'high_demand_items': df.nlargest(10, 'views')[['title', 'price', 'views']]
    }
    
    return price_analysis, demand_analysis

def create_market_report(price_analysis, demand_analysis):
    """Generate market analysis report"""
    
    report = f"""
    WOODWORKING MARKET ANALYSIS REPORT
    ================================
    
    PRICING INSIGHTS:
    - Average Price: ${price_analysis['avg_price']:.2f}
    - Median Price: ${price_analysis['median_price']:.2f}
    - Price Range (25th-75th percentile): ${price_analysis['price_range']['low']:.2f} - ${price_analysis['price_range']['high']:.2f}
    
    DEMAND INDICATORS:
    - Average Views per Item: {demand_analysis['avg_views']:.0f}
    - Average Favorites per Item: {demand_analysis['avg_favorites']:.0f}
    
    TOP PERFORMING ITEMS:
    {demand_analysis['high_demand_items'].to_string()}
    """
    
    return report

ModuleNotFoundError: No module named 'matplotlib'

## Web Scraping for Public Business Directories

In [ ]:
# Scraping Yellow Pages or similar business directories (public data)
import requests
from bs4 import BeautifulSoup
import time

def scrape_business_directory(location, category="furniture"):
    """Scrape public business directory data"""
    
    # This is for publicly available directory information
    base_url = "https://www.yellowpages.com/search"
    
    params = {
        'search_terms': f'{category} woodworking',
        'geo_location_terms': location
    }
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    
    businesses = []
    
    try:
        response = requests.get(base_url, params=params, headers=headers)
        soup = BeautifulSoup(response.content, 'html.parser')
        
        # Extract business information (adjust selectors based on site structure)
        business_listings = soup.find_all('div', class_='result')
        
        for listing in business_listings:
            name = listing.find('a', class_='business-name')
            rating = listing.find('div', class_='rating')
            
            if name:
                businesses.append({
                    'name': name.get_text().strip(),
                    'rating': rating.get_text().strip() if rating else 'N/A'
                })
        
        time.sleep(2)  # Be respectful with requests
        
    except Exception as e:
        print(f"Error scraping directory: {e}")
    
    return businesses

## Social Media Market Research

In [ ]:
# Using official APIs for market research
def analyze_woodworking_social_trends():
    """Analyze social media trends for woodworking"""
    
    # Instagram Hashtag Analysis (using official API)
    woodworking_hashtags = [
        '#customfurniture', '#woodworking', '#handmade', 
        '#liveedge', '#woodwork', '#furniture'
    ]
    
    # YouTube Data API for woodworking content analysis
    def get_youtube_woodworking_data(api_key):
        from googleapiclient.discovery import build
        
        youtube = build('youtube', 'v3', developerKey=api_key)
        
        search_response = youtube.search().list(
            q='woodworking business custom furniture',
            part='id,snippet',
            maxResults=50,
            type='video',
            order='relevance'
        ).execute()
        
        video_data = []
        for item in search_response['items']:
            video_data.append({
                'title': item['snippet']['title'],
                'description': item['snippet']['description'],
                'published_at': item['snippet']['publishedAt']
            })
        
        return video_data
    
    return woodworking_hashtags

## Demand Forecasting Model

In [ ]:
import pandas as pd
from sklearn.linear_model import LinearRegression
import numpy as np

def create_demand_forecast(historical_data):
    """Create demand forecast based on collected data"""
    
    df = pd.DataFrame(historical_data)
    df['date'] = pd.to_datetime(df['date'])
    df['month'] = df['date'].dt.month
    df['year'] = df['date'].dt.year
    
    # Features for demand prediction
    features = ['month', 'year', 'avg_price', 'search_volume']
    X = df[features]
    y = df['demand_indicator']  # Views, sales, etc.
    
    # Simple linear regression model
    model = LinearRegression()
    model.fit(X, y)
    
    # Forecast next 6 months
    future_months = []
    current_date = pd.Timestamp.now()
    
    for i in range(1, 7):
        future_date = current_date + pd.DateOffset(months=i)
        future_months.append({
            'month': future_date.month,
            'year': future_date.year,
            'avg_price': df['avg_price'].mean(),  # Use historical average
            'search_volume': df['search_volume'].mean()
        })
    
    future_df = pd.DataFrame(future_months)
    forecast = model.predict(future_df[features])
    
    return forecast, future_df

def generate_business_insights(all_data):
    """Generate actionable business insights"""
    
    insights = {
        'optimal_price_range': None,
        'high_demand_products': [],
        'seasonal_trends': {},
        'competitor_gaps': [],
        'market_opportunities': []
    }
    
    # Analyze optimal pricing
    price_demand_correlation = np.corrcoef(all_data['prices'], all_data['demand'])
    
    # Identify market gaps
    # Analyze competitor offerings vs. market demand
    
    return insights

## Data Collection Automation

In [ ]:
import schedule
import time
from datetime import datetime

class WoodworkingMarketResearch:
    def __init__(self):
        self.data_store = []
    
    def daily_data_collection(self):
        """Collect market data daily"""
        
        # Google Trends data
        trends_data = analyze_woodworking_trends()
        
        # Local business data
        local_data = analyze_local_woodworking_businesses("Your City, State", "API_KEY")
        
        # E-commerce data
        etsy_data = analyze_etsy_woodworking_market("ETSY_API_KEY")
        
        # Combine and store data
        daily_snapshot = {
            'date': datetime.now(),
            'trends': trends_data,
            'local_businesses': local_data,
            'ecommerce_data': etsy_data
        }
        
        self.data_store.append(daily_snapshot)
        
        # Save to CSV
        self.save_daily_report()
    
    def save_daily_report(self):
        """Save collected data to CSV"""
        if self.data_store:
            df = pd.DataFrame(self.data_store)
            filename = f"woodworking_market_data_{datetime.now().strftime('%Y%m%d')}.csv"
            df.to_csv(filename, index=False)
            print(f"Market data saved: {filename}")

# Schedule daily data collection
researcher = WoodworkingMarketResearch()
schedule.every().day.at("09:00").do(researcher.daily_data_collection)

# Run the scheduler
while True:
    schedule.run_pending()
    time.sleep(3600)  # Check every hour

Marketing and Forecasting Research
=====================================

# Market Research Data Sources

## Industry Reports & Analytics

# Google Trends for woodworking demand
from pytrends.request import TrendReq

def analyze_woodworking_trends():
    pytrends = TrendReq(hl='en-US', tz=360)
    
    # Woodworking-related keywords
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
    
    # Get interest over time
    trends_data = pytrends.interest_over_time()
    
    # Get related queries
    related_queries = pytrends.related_queries()
    
    return trends_data, related_queries

trends, queries = analyze_woodworking_trends()

### Anthropic Attempted Fix

import time
import random
from pytrends.request import TrendReq
import pandas as pd

def analyze_woodworking_trends():
    try:
        # Initialize with different settings
        pytrends = TrendReq(hl='en-US', tz=360, timeout=(10,25), retries=2, backoff_factor=0.1)
        
        # Use fewer keywords (max 5 per request)
        keywords = ['custom furniture', 'woodworking', 'handmade furniture']
        
        # Add random delay to avoid rate limiting
        time.sleep(random.uniform(1, 3))
        
        print("Building payload...")
        pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
        
        print("Getting interest over time...")
        trends_data = pytrends.interest_over_time()
        
        # Add delay before next request
        time.sleep(random.uniform(2, 4))
        
        print("Getting related queries...")
        related_queries = pytrends.related_queries()
        
        return trends_data, related_queries
        
    except Exception as e:
        print(f"Error: {e}")
        print("Trying alternative approach...")
        return analyze_with_smaller_batches()

def analyze_with_smaller_batches():
    """Alternative approach with individual keyword analysis"""
    pytrends = TrendReq(hl='en-US', tz=360, timeout=(10,25), retries=2, backoff_factor=0.1)
    
    keywords = ['woodworking', 'custom furniture', 'handmade furniture']
    all_data = {}
    
    for keyword in keywords:
        try:
            print(f"Analyzing: {keyword}")
            time.sleep(random.uniform(2, 4))  # Longer delay between requests
            
            pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
            trends = pytrends.interest_over_time()
            
            if not trends.empty:
                all_data[keyword] = trends[keyword]
            
        except Exception as e:
            print(f"Failed for {keyword}: {e}")
            continue
    
    if all_data:
        combined_df = pd.DataFrame(all_data)
        return combined_df, None
    else:
        return None, None

# Run the analysis
print("Starting woodworking trends analysis...")
trends, queries = analyze_woodworking_trends()

if trends is not None and not trends.empty:
    print("\n=== TRENDS DATA ===")
    print(trends.head())
    print(f"\nShape: {trends.shape}")
    
    if queries:
        print("\n=== RELATED QUERIES ===")
        for keyword, query_data in queries.items():
            print(f"\n{keyword.upper()}:")
            if query_data['top'] is not None:
                print("Top queries:")
                print(query_data['top'].head())
else:
    print("No data retrieved. Google Trends might be blocking requests.")
    print("Try again later or use a different approach.")

### Google Fix

# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time # Import the time module to pause between requests

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    Fetches interest over time for a list of keywords and then fetches
    related queries for each keyword individually.

    Returns:
        pd.DataFrame: A DataFrame containing interest over time data.
        dict: A dictionary containing related queries for each keyword.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    # Woodworking-related keywords
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time for all keywords (for comparison) ---
    print("Fetching interest over time for all keywords...")
    pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
    trends_data = pytrends.interest_over_time()
    
    # --- 2. Get Related Queries for each keyword individually ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        # Build a new payload for each individual keyword
        pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
        
        # Add a small delay to avoid making requests too quickly
        time.sleep(1) 
        
        try:
            related_queries_dict[keyword] = pytrends.related_queries()
        except Exception as e:
            print(f"Could not fetch related queries for '{keyword}'. Error: {e}")
            related_queries_dict[keyword] = None

    return trends_data, related_queries_dict

# Run the analysis
trends_df, related_queries = analyze_woodworking_trends()

# --- Display the results ---
print("\n--- Interest Over Time ---")
print(trends_df.tail()) # Print the last 5 rows of the trend data

print("\n\n--- Related Queries ---")
for keyword, queries in related_queries.items():
    if queries:
        print(f"\nTop Related Queries for '{keyword}':")
        # The result is a dictionary where the value is another dictionary with 'top' and 'rising'
        top_queries = queries[keyword]['top']
        if top_queries is not None and not top_queries.empty:
            print(top_queries)
        else:
            print("No top related queries found.")

#### Google Response to Broken API Endpoint

     ```

# Google Trends for woodworking demand

from pytrends.request import TrendReq


def analyze_woodworking_trends():

pytrends = TrendReq(hl='en-US', tz=360)

# Woodworking-related keywords

keywords = ['custom furniture', 'woodworking', 'handmade furniture',

'live edge table', 'custom cabinets']

pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')

# Get interest over time

trends_data = pytrends.interest_over_time()

# Get related queries

related_queries = pytrends.related_queries()

return trends_data, related_queries


trends, queries = analyze_woodworking_trends()

```


/home/lubuntu/myenv/lib/python3.12/site-packages/pytrends/request.py:260: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`

df = df.fillna(False)


---------------------------------------------------------------------------

IndexError Traceback (most recent call last)

Cell In[4], line 21

17 related_queries = pytrends.related_queries()

19 return trends_data, related_queries

---> 21 trends, queries = analyze_woodworking_trends()

Cell In[4], line 17, in analyze_woodworking_trends()

14 trends_data = pytrends.interest_over_time()

16 # Get related queries

---> 17 related_queries = pytrends.related_queries()

19 return trends_data, related_queries

File ~/myenv/lib/python3.12/site-packages/pytrends/request.py:450, in TrendReq.related_queries(self)

447 # top queries

448 try:

449 top_df = pd.DataFrame(

--> 450 req_json['default']['rankedList'][0]['rankedKeyword'])

451 top_df = top_df[['query', 'value']]

452 except KeyError:

453 # in case no top queries are found, the lines above will throw a KeyError

IndexError: list index out of range

Outline

    Diagnosis of the IndexError: An explanation of why the IndexError: list index out of range occurs when using related_queries() with multiple keywords.

    The Solution: A description of the correct approach, which involves iterating through each keyword to fetch its related queries individually.

    Corrected Python Code: A complete, runnable code snippet that implements the solution and correctly fetches both interest over time for all keywords and related queries for each keyword.

    Explanation of Changes: A step-by-step breakdown of the modifications made to the original code.

    A Note on the FutureWarning: A brief explanation of the warning message appearing in the output.

1. Diagnosis of the IndexError

Hello! I can help you with this Python error. The IndexError: list index out of range you're encountering is a common issue when using the pytrends library.

The root cause of the problem is that the pytrends.related_queries() method is designed to fetch data for only one keyword at a time.

In your original code, you build a single payload with a list of five keywords:
pytrends.build_payload(keywords, ...)

While interest_over_time() can handle multiple keywords to generate a comparison chart, related_queries() cannot. When it receives a payload with more than one keyword, the Google Trends API returns a data structure that pytrends doesn't expect. Specifically, the list that should contain the related query data (rankedList) comes back empty. The library then tries to access the first element of this empty list ([0]), which triggers the IndexError.

2. The Solution

To fix this, you need to separate the logic for fetching "interest over time" and "related queries."

    Interest Over Time: You can continue to fetch this data in a single call with all your keywords to get a comparison.

    Related Queries: You must loop through your list of keywords and make a separate API call for each one. You can store the results in a dictionary where the keys are your keywords and the values are the corresponding related queries.

3. Corrected Python Code

Here is the corrected and complete code. This version first gets the comparison data for all keywords and then iterates through each keyword to safely get its related queries.
Python

# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time # Import the time module to pause between requests

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    Fetches interest over time for a list of keywords and then fetches
    related queries for each keyword individually.

    Returns:
        pd.DataFrame: A DataFrame containing interest over time data.
        dict: A dictionary containing related queries for each keyword.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    # Woodworking-related keywords
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time for all keywords (for comparison) ---
    print("Fetching interest over time for all keywords...")
    pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
    trends_data = pytrends.interest_over_time()
    
    # --- 2. Get Related Queries for each keyword individually ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        # Build a new payload for each individual keyword
        pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
        
        # Add a small delay to avoid making requests too quickly
        time.sleep(1) 
        
        try:
            related_queries_dict[keyword] = pytrends.related_queries()
        except Exception as e:
            print(f"Could not fetch related queries for '{keyword}'. Error: {e}")
            related_queries_dict[keyword] = None

    return trends_data, related_queries_dict

# Run the analysis
trends_df, related_queries = analyze_woodworking_trends()

# --- Display the results ---
print("\n--- Interest Over Time ---")
print(trends_df.tail()) # Print the last 5 rows of the trend data

print("\n\n--- Related Queries ---")
for keyword, queries in related_queries.items():
    if queries:
        print(f"\nTop Related Queries for '{keyword}':")
        # The result is a dictionary where the value is another dictionary with 'top' and 'rising'
        top_queries = queries[keyword]['top']
        if top_queries is not None and not top_queries.empty:
            print(top_queries)
        else:
            print("No top related queries found.")

4. Explanation of Changes

    Separate API Calls: The core logic is now split. The call to get interest_over_time remains as it was, using the full list of keywords.

    Loop for Related Queries: A for loop iterates through your keywords list.

    Single-Keyword Payload: Inside the loop, pytrends.build_payload([keyword], ...) is called. Note that keyword is enclosed in square brackets [] because the method expects a list, even if it's just a list containing one item.

    Storing Results: The results from pytrends.related_queries() are stored in a dictionary called related_queries_dict. This is a clean way to associate each keyword with its specific results.

    Error Handling & Delay:

        time.sleep(1): A one-second pause is added in the loop. It's good practice to add a small delay between API calls to avoid being rate-limited or blocked by Google for making too many requests too quickly.

        try...except: I've wrapped the related_queries() call in a try...except block. This makes the code more robust. If Google Trends has no related query data for a specific term, it might still cause an error, and this block will catch it, print a message, and allow the script to continue to the next keyword instead of crashing.

5. A Note on the FutureWarning

You may still see this warning in your output:
FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated...

This warning is not causing your IndexError. It originates from the pandas library, which pytrends uses internally. It's simply informing you that a function pandas uses will behave differently in a future version. This is a message for the developers of pytrends to address in a future update. You can safely ignore this warning as it does not affect the execution or results of your code.


Fetching related queries for each keyword...

Fetching for: 'custom furniture'

Could not fetch related queries for 'custom furniture'. Error: list index out of range

Fetching for: 'woodworking'

Could not fetch related queries for 'woodworking'. Error: list index out of range

Fetching for: 'handmade furniture'

Could not fetch related queries for 'handmade furniture'. Error: list index out of range

Fetching for: 'live edge table'

Could not fetch related queries for 'live edge table'. Error: list index out of range

Fetching for: 'custom cabinets'

Could not fetch related queries for 'custom cabinets'. Error: list index out of range


--- Interest Over Time ---

custom furniture woodworking handmade furniture \

date

2025-06-01 10 70 1

2025-06-08 12 68 1

2025-06-15 10 71 1

2025-06-22 10 73 1

2025-06-29 11 66 1


live edge table custom cabinets isPartial

date

2025-06-01 5 15 False

2025-06-08 4 15 False

2025-06-15 5 15 False

2025-06-22 5 15 False

2025-06-29 5 13 True



--- Related Queries ---


You've run into a frustrating and unfortunately common issue with the pytrends library. Your new output confirms that the error isn't just about sending multiple keywords at once, but is a more fundamental problem with the related_queries() function itself.

Let's break down what's happening and how to build a more robust script to handle it.

Outline

    Diagnosis of the Persistent Error: An explanation of why the IndexError continues to happen even with single keywords, pointing to an issue within the pytrends library itself.

    The Fragility of Unofficial Wrappers: A brief look at why libraries like pytrends can sometimes break.

    Robust Solution: A revised code block that anticipates these errors and safely handles the data that is returned, preventing crashes and extracting whatever information is available.

    Interpreting the New Results: How to understand the output from the robust script.

1. Diagnosis of the Persistent Error

The fact that you are now getting the IndexError: list index out of range for every single keyword tells us the problem lies within the pytrends library's ability to parse the response it's getting from Google.

Here's the situation:

    You send a valid request for a single keyword (e.g., 'woodworking').

    Google Trends sends back a response.

    The structure of this response has likely changed slightly from what the pytrends library was originally designed to expect.

    When the related_queries() method tries to find the data in the specific location it's hard-coded to look (req_json['default']['rankedList'][0]), it finds an empty list ([]) and crashes when it tries to get the first item ([0]).

This happens even when there are related queries visible on the Google Trends website. The issue is purely in the parsing step within the library.

2. The Fragility of Unofficial Wrappers

pytrends is an unofficial "wrapper" library. It works by mimicking a web browser to scrape the Google Trends website and access its internal API. This is a brilliant but fragile approach. If Google changes the backend structure of their website or API—even slightly—the library can break until its developers release a new version that accounts for the change. You appear to have caught it during one of these broken periods.

3. Robust Solution

We cannot fix the pytrends library itself, but we can write our Python code to be more defensive. The goal is to successfully call the function and then carefully inspect whatever data we get back, rather than assuming it's in the correct format.

The code below is designed to handle this IndexError gracefully. It attempts the call and then safely checks the structure of the returned dictionary before trying to extract the 'top' and 'rising' queries.
Python

# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    This robust version anticipates potential errors from the pytrends library
    and handles them gracefully.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time (this part usually works reliably) ---
    print("Fetching interest over time for all keywords...")
    try:
        pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
        trends_data = pytrends.interest_over_time()
    except Exception as e:
        print(f"Could not fetch interest over time. Error: {e}")
        trends_data = None
    
    # --- 2. Get Related Queries with Robust Error Handling ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        time.sleep(2) # Increased delay to be safe
        
        try:
            # Re-build payload for each keyword
            pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
            related_queries_dict[keyword] = pytrends.related_queries()
        except IndexError:
            # This is the specific error we are seeing.
            print(f"  > Caught IndexError. Google's API response is not in the expected format for '{keyword}'.")
            related_queries_dict[keyword] = None # Mark as failed
        except Exception as e:
            # Catch any other potential errors
            print(f"  > An unexpected error occurred for '{keyword}': {e}")
            related_queries_dict[keyword] = None # Mark as failed
            
    return trends_data, related_queries_dict

# --- Run the analysis and display results safely ---
trends_df, related_queries = analyze_woodworking_trends()

if trends_df is not None:
    print("\n\n--- Interest Over Time ---")
    print(trends_df.tail())
else:
    print("\n\n--- No Interest Over Time data was fetched ---")

print("\n\n--- Related Queries ---")
if not related_queries:
    print("No related queries could be fetched.")
else:
    for keyword, result in related_queries.items():
        print(f"\n--- Results for '{keyword}' ---")
        if result is None:
            print("  > No data returned due to an API parsing error.")
            continue
        
        # Safely access 'top' and 'rising' queries
        top_queries = result.get(keyword, {}).get('top')
        rising_queries = result.get(keyword, {}).get('rising')
        
        if top_queries is not None and not top_queries.empty:
            print("\n  Top Related Queries:")
            print(top_queries)
        else:
            print("\n  > No 'Top' related queries were found or could be parsed.")
            
        if rising_queries is not None and not rising_queries.empty:
            print("\n  Rising Related Queries:")
            print(rising_queries)
        else:
            print("\n  > No 'Rising' related queries were found or could be parsed.")

4. Interpreting the New Results

When you run this new script, you will likely see a similar output to what you got before, but with more descriptive messages:
Plaintext

Fetching for: 'custom furniture'
  > Caught IndexError. Google's API response is not in the expected format for 'custom furniture'.
... (and so on for other keywords)

At the end, the results section will clearly state that no data could be parsed.

What to do now?

    Wait for an Update: The most common solution is to wait for the pytrends developers to patch the library. You can check for updates periodically by running pip install --upgrade pytrends.

    Try Different Parameters: You can experiment by changing the timeframe. Sometimes a shorter timeframe like 'today 3-m' or 'now 7-d' might use a different API endpoint that isn't broken.

    Use the Website: For the time being, the most reliable way to get related queries is to go directly to the Google Trends website and search for your terms manually.



## E-commerce Platform APIs

# Google Trends for woodworking demand
import pandas as pd
from pytrends.request import TrendReq
import time

def analyze_woodworking_trends():
    """
    Analyzes Google Trends for woodworking keywords.

    This robust version anticipates potential errors from the pytrends library
    and handles them gracefully.
    """
    pytrends = TrendReq(hl='en-US', tz=360)
    
    keywords = ['custom furniture', 'woodworking', 'handmade furniture', 
               'live edge table', 'custom cabinets']
    
    # --- 1. Get Interest Over Time (this part usually works reliably) ---
    print("Fetching interest over time for all keywords...")
    try:
        pytrends.build_payload(keywords, timeframe='today 12-m', geo='US')
        trends_data = pytrends.interest_over_time()
    except Exception as e:
        print(f"Could not fetch interest over time. Error: {e}")
        trends_data = None
    
    # --- 2. Get Related Queries with Robust Error Handling ---
    related_queries_dict = {}
    print("\nFetching related queries for each keyword...")
    
    for keyword in keywords:
        print(f"Fetching for: '{keyword}'")
        time.sleep(2) # Increased delay to be safe
        
        try:
            # Re-build payload for each keyword
            pytrends.build_payload([keyword], timeframe='today 12-m', geo='US')
            related_queries_dict[keyword] = pytrends.related_queries()
        except IndexError:
            # This is the specific error we are seeing.
            print(f"  > Caught IndexError. Google's API response is not in the expected format for '{keyword}'.")
            related_queries_dict[keyword] = None # Mark as failed
        except Exception as e:
            # Catch any other potential errors
            print(f"  > An unexpected error occurred for '{keyword}': {e}")
            related_queries_dict[keyword] = None # Mark as failed
            
    return trends_data, related_queries_dict

# --- Run the analysis and display results safely ---
trends_df, related_queries = analyze_woodworking_trends()

if trends_df is not None:
    print("\n\n--- Interest Over Time ---")
    print(trends_df.tail())
else:
    print("\n\n--- No Interest Over Time data was fetched ---")

print("\n\n--- Related Queries ---")
if not related_queries:
    print("No related queries could be fetched.")
else:
    for keyword, result in related_queries.items():
        print(f"\n--- Results for '{keyword}' ---")
        if result is None:
            print("  > No data returned due to an API parsing error.")
            continue
        
        # Safely access 'top' and 'rising' queries
        top_queries = result.get(keyword, {}).get('top')
        rising_queries = result.get(keyword, {}).get('rising')
        
        if top_queries is not None and not top_queries.empty:
            print("\n  Top Related Queries:")
            print(top_queries)
        else:
            print("\n  > No 'Top' related queries were found or could be parsed.")
            
        if rising_queries is not None and not rising_queries.empty:
            print("\n  Rising Related Queries:")
            print(rising_queries)
        else:
            print("\n  > No 'Rising' related queries were found or could be parsed.")

# Etsy API for handmade/custom furniture market
import requests

def analyze_etsy_woodworking_market(api_key):
    """Analyze woodworking market on Etsy"""
    
    # Search for woodworking items
    url = "https://openapi.etsy.com/v3/application/listings/active"
    
    headers = {
        'x-api-key': api_key,
        'Content-Type': 'application/json'
    }
    
    params = {
        'keywords': 'custom furniture woodworking',
        'limit': 100,
        'includes': 'Images,Shop,User',
        'sort_on': 'score'
    }
    
    response = requests.get(url, headers=headers, params=params)
    listings = response.json()
    
    # Analyze pricing and demand indicators
    market_data = []
    for listing in listings.get('results', []):
        market_data.append({
            'title': listing.get('title'),
            'price': listing.get('price', {}).get('amount'),
            'currency': listing.get('price', {}).get('currency_code'),
            'views': listing.get('views'),
            'num_favorers': listing.get('num_favorers'),
            'shop_name': listing.get('Shop', {}).get('shop_name'),
            'creation_timestamp': listing.get('creation_timestamp')
        })
    
    return market_data

## Local Business Intelligence

# Google My Business API for local competition analysis
def analyze_local_woodworking_businesses(location, gmb_api_key):
    """Find local woodworking businesses and their performance"""
    import googlemaps
    
    gmaps = googlemaps.Client(key=gmb_api_key)
    
    # Search for woodworking businesses
    search_terms = [
        'custom furniture',
        'woodworking shop',
        'cabinet maker',
        'furniture maker'
    ]
    
    local_businesses = []
    
    for term in search_terms:
        places = gmaps.places_nearby(
            location=location,
            radius=50000,  # 50km radius
            keyword=term,
            type='establishment'
        )
        
        for place in places.get('results', []):
            place_details = gmaps.place(
                place_id=place['place_id'],
                fields=['name', 'rating', 'user_ratings_total', 'website', 
                       'formatted_phone_number', 'opening_hours', 'photos']
            )
            
            local_businesses.append({
                'name': place_details['result']['name'],
                'rating': place_details['result'].get('rating'),
                'review_count': place_details['result'].get('user_ratings_total'),
                'website': place_details['result'].get('website'),
                'search_term': term
            })
    
    return local_businesses

## Price Analysis Tools



import pandas as pd
import matplotlib.pyplot as plt

def analyze_pricing_data(market_data):
    """Analyze pricing patterns in woodworking market"""
    
    df = pd.DataFrame(market_data)
    
    # Price analysis by category
    price_analysis = {
        'avg_price': df['price'].mean(),
        'median_price': df['price'].median(),
        'price_range': {
            'low': df['price'].quantile(0.25),
            'high': df['price'].quantile(0.75)
        },
        'popular_price_points': df['price'].value_counts().head(10)
    }
    
    # Demand indicators
    demand_analysis = {
        'avg_views': df['views'].mean(),
        'avg_favorites': df['num_favorers'].mean(),
        'high_demand_items': df.nlargest(10, 'views')[['title', 'price', 'views']]
    }
    
    return price_analysis, demand_analysis

def create_market_report(price_analysis, demand_analysis):
    """Generate market analysis report"""
    
    report = f"""
    WOODWORKING MARKET ANALYSIS REPORT
    ================================
    
    PRICING INSIGHTS:
    - Average Price: ${price_analysis['avg_price']:.2f}
    - Median Price: ${price_analysis['median_price']:.2f}
    - Price Range (25th-75th percentile): ${price_analysis['price_range']['low']:.2f} - ${price_analysis['price_range']['high']:.2f}
    
    DEMAND INDICATORS:
    - Average Views per Item: {demand_analysis['avg_views']:.0f}
    - Average Favorites per Item: {demand_analysis['avg_favorites']:.0f}
    
    TOP PERFORMING ITEMS:
    {demand_analysis['high_demand_items'].to_string()}
    """
    
    return report

## Web Scraping for Public Business Directories

# Scraping Yellow Pages or similar business directories (public data)
import requests
from bs4 import BeautifulSoup
import time

def scrape_business_directory(location, category="furniture"):
    """Scrape public business directory data"""
    
    # This is for publicly available directory information
    base_url = "https://www.yellowpages.com/search"
    
    params = {
        'search_terms': f'{category} woodworking',
        'geo_location_terms': location
    }
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    
    businesses = []
    
    try:
        response = requests.get(base_url, params=params, headers=headers)
        soup = BeautifulSoup(response.content, 'html.parser')
        
        # Extract business information (adjust selectors based on site structure)
        business_listings = soup.find_all('div', class_='result')
        
        for listing in business_listings:
            name = listing.find('a', class_='business-name')
            rating = listing.find('div', class_='rating')
            
            if name:
                businesses.append({
                    'name': name.get_text().strip(),
                    'rating': rating.get_text().strip() if rating else 'N/A'
                })
        
        time.sleep(2)  # Be respectful with requests
        
    except Exception as e:
        print(f"Error scraping directory: {e}")
    
    return businesses

## Social Media Market Research

# Using official APIs for market research
def analyze_woodworking_social_trends():
    """Analyze social media trends for woodworking"""
    
    # Instagram Hashtag Analysis (using official API)
    woodworking_hashtags = [
        '#customfurniture', '#woodworking', '#handmade', 
        '#liveedge', '#woodwork', '#furniture'
    ]
    
    # YouTube Data API for woodworking content analysis
    def get_youtube_woodworking_data(api_key):
        from googleapiclient.discovery import build
        
        youtube = build('youtube', 'v3', developerKey=api_key)
        
        search_response = youtube.search().list(
            q='woodworking business custom furniture',
            part='id,snippet',
            maxResults=50,
            type='video',
            order='relevance'
        ).execute()
        
        video_data = []
        for item in search_response['items']:
            video_data.append({
                'title': item['snippet']['title'],
                'description': item['snippet']['description'],
                'published_at': item['snippet']['publishedAt']
            })
        
        return video_data
    
    return woodworking_hashtags

## Demand Forecasting Model

import pandas as pd
from sklearn.linear_model import LinearRegression
import numpy as np

def create_demand_forecast(historical_data):
    """Create demand forecast based on collected data"""
    
    df = pd.DataFrame(historical_data)
    df['date'] = pd.to_datetime(df['date'])
    df['month'] = df['date'].dt.month
    df['year'] = df['date'].dt.year
    
    # Features for demand prediction
    features = ['month', 'year', 'avg_price', 'search_volume']
    X = df[features]
    y = df['demand_indicator']  # Views, sales, etc.
    
    # Simple linear regression model
    model = LinearRegression()
    model.fit(X, y)
    
    # Forecast next 6 months
    future_months = []
    current_date = pd.Timestamp.now()
    
    for i in range(1, 7):
        future_date = current_date + pd.DateOffset(months=i)
        future_months.append({
            'month': future_date.month,
            'year': future_date.year,
            'avg_price': df['avg_price'].mean(),  # Use historical average
            'search_volume': df['search_volume'].mean()
        })
    
    future_df = pd.DataFrame(future_months)
    forecast = model.predict(future_df[features])
    
    return forecast, future_df

def generate_business_insights(all_data):
    """Generate actionable business insights"""
    
    insights = {
        'optimal_price_range': None,
        'high_demand_products': [],
        'seasonal_trends': {},
        'competitor_gaps': [],
        'market_opportunities': []
    }
    
    # Analyze optimal pricing
    price_demand_correlation = np.corrcoef(all_data['prices'], all_data['demand'])
    
    # Identify market gaps
    # Analyze competitor offerings vs. market demand
    
    return insights

## Data Collection Automation

import schedule
import time
from datetime import datetime

class WoodworkingMarketResearch:
    def __init__(self):
        self.data_store = []
    
    def daily_data_collection(self):
        """Collect market data daily"""
        
        # Google Trends data
        trends_data = analyze_woodworking_trends()
        
        # Local business data
        local_data = analyze_local_woodworking_businesses("Your City, State", "API_KEY")
        
        # E-commerce data
        etsy_data = analyze_etsy_woodworking_market("ETSY_API_KEY")
        
        # Combine and store data
        daily_snapshot = {
            'date': datetime.now(),
            'trends': trends_data,
            'local_businesses': local_data,
            'ecommerce_data': etsy_data
        }
        
        self.data_store.append(daily_snapshot)
        
        # Save to CSV
        self.save_daily_report()
    
    def save_daily_report(self):
        """Save collected data to CSV"""
        if self.data_store:
            df = pd.DataFrame(self.data_store)
            filename = f"woodworking_market_data_{datetime.now().strftime('%Y%m%d')}.csv"
            df.to_csv(filename, index=False)
            print(f"Market data saved: {filename}")

# Schedule daily data collection
researcher = WoodworkingMarketResearch()
schedule.every().day.at("09:00").do(researcher.daily_data_collection)

# Run the scheduler
while True:
    schedule.run_pending()
    time.sleep(3600)  # Check every hour
